### DISCUSSION TITLE
pocketbase/tests with inmemory sqlite

### DISCUSSION THREAD

**Discussion by lyuboxa** [pocketbase/pocketbase#3444] [Q&A] [answered]
Hello folks,

The pocketbase tests tooling doesn't seem to support  the use of `inmemory` sqlite, 
which can decrease the cost for smaller tests. I found a previous discussion on the topic which somewhat vaguely suggests
that it may not be a good idea to do this.

Any particular reason why this may not work?

Thanks
Lyubo

**Comment by ganigeorgiev** [comment]
The tests should work fine if an inmemory db was used.

The main reason we don't do it is because it complicates the internals and the tests don't really take that much time to run.
It is also a lot easier to maintain because we can directly use the [test directory](https://github.com/pocketbase/pocketbase/tree/master/tests/data) as "fixtures".
Or in other words, it allows me to browse and work with the test db just like a regular PocketBase installation using something like:
```sh
./pocketbase serve --dir="/path/to/test"
```

_It may not be the best setup, and maybe instead of copying the entire db we can have `.json` files for each table that we can import on every test run, as this will allow more easier review when multiple devs updates the tests db, but for now I don't want to complicate it unnecessary._

---

_Side-note: there are plans to allow more "native" SQLite dev experience in the future, which will allow devs to define their own db initialization setup (eg. to load extensions or completely custom SQLite driver)._

_Edit: https://pocketbase.io/docs/go-overview/#custom-sqlite-driver_

**Reply by lyuboxa** [reply]
Thanks, appreciate your input.

I agree, in some cases using seeded fixtures can be all that's needed.
This definitely reduces the IO cost associated with bootstrapping a new database file, migrations, etc.

In some cases in order to allow for code to be unit tested, e.g. `daos.Dao` will need be isolated using interface contracts
which essentially wraps record/collection methods. I somewhat dislike this method as it creates unnecessary layers and would rather make testing via `tests` as cheaper as possible, and the low hanging fruit here is IO related. 

From what I can tell, the bootstrapping and configuration does not allow for file paths to be specified directly to dbx/sqlite, 
will you be open to allow a little bit more flexibility in that area? To specifically allow to either: allow in-memory storage; or way open up the configuration to point to a file directly and recognize `file::memory`/`:memory` as valid choice.

Thanks
Lyubo

**Reply by ganigeorgiev** [reply]
> will you be open to allow a little bit more flexibility in that area

Yes, that was what I meant by _"allow devs to define their own db initialization setup"_. In other words, eventually in the future there will be an option to specify a custom [connectDB function](https://github.com/pocketbase/pocketbase/blob/7fa1ff53c9bc7c535a89d78ba491c18987e3bb48/core/db_cgo.go#L43), but for now this remains a low priority as I'm planning an internal refactoring somewhere after v0.20 (_see also https://github.com/pocketbase/pocketbase/discussions/3271_).

**Reply by lyuboxa** [reply]
Sorry, should have been clear, I can take a stab at implementing this and let you decide where you want to take it. Happy to do it.

**Reply by ganigeorgiev** [reply, marked as answer]
I'm not sure what the public API should look like at the moment and most likely it will change with the planned refactoring, so it may be better to wait.

If the above represents a blocker issue for your use case, feel free to file a PR but there are no promises that it will get merged or reviewed in a timely manner.

**Reply by lyuboxa** [reply]
Sounds good and makes sense, thanks for response.

**Comment by VittorioParagallo** [comment]
Hello are there any news on this topic expecially on the possibility fo inmemory db? i'm coding in dart and now that i upgraded to v23 i started thinking that would be more safe to write the unit testing for my db calls to ease future pb upgrades, any suggestion? i've been reading a bit around but haven't figure out a method to populate in the beginning of the test and not worring to clean at the end of the test, that's why to reduce boilerplate i was thinking about the possibility to load the test db in memory. Thank in advance for the help

**Reply by ganigeorgiev** [reply]
@VittorioParagallo It is not clear what problem you are trying to solve, nor how it is related to Dart or being able to open an in-memory db, but nothing from the above discussion has changed in regards to [how PocketBase runs its automated tests](https://pocketbase.io/docs/go-testing/).

With v0.23 you can specify your own custom dsn string or completely different SQLite driver, see https://pocketbase.io/docs/go-overview/#custom-sqlite-driver, but I'd advice against doing that unless you have clear understanding on the problem that you are trying to solve.
